File: flutter/android/src/main/kotlin/com/spruceid/sprucekit_mobile/DcApiAdapter.kt
package com.spruceid.sprucekit_mobile

import android.app.Application
import android.content.Context
import android.os.Build
import android.util.Log
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.StorageManager
import com.spruceid.mobile.sdk.dcapi.Registry
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap

/**
 * DC API Adapter for Android
 *
 * Handles registering mDoc credentials with Android CredentialManager
 * for the Digital Credentials API.
 */
internal class DcApiAdapter(
    private val context: Context,
    private val credentialPackAdapter: CredentialPackAdapter
) : DcApi {

    private val application: Application
        get() = context.applicationContext as Application

    private val registeredCredentials = ConcurrentHashMap<String, RegisteredCredentialInfo>()
    private var registry: Registry? = null
    private val storageManager by lazy { StorageManager(context) }

    companion object {
        private const val TAG = "DcApiAdapter"
        // Icon file name in the app's assets folder (e.g., "icon.ico")
        // Apps must provide this file in: android/app/src/main/assets/
        private const val DEFAULT_ICON = "icon.ico"
    }

    override fun syncCredentialsToAppGroup(
        appGroupId: String,
        packIds: List<String>,
        callback: (Result<DcApiResult>) -> Unit
    ) {
        // On Android, we don't use App Groups like iOS
        // Instead, CredentialManager handles credential storage and access
        // This method is a no-op on Android but returns success for API compatibility
        callback(Result.success(DcApiSuccess(
            message = "App Group sync not needed on Android (handled by CredentialManager)"
        )))
    }

    override fun registerCredentials(
        packIds: List<String>,
        walletName: String?,
        callback: (Result<DcApiResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Initialize registry
                // Note: walletName parameter reserved for future use when SDK supports it
                registry = Registry(application, DEFAULT_ICON)

                // Collect all credential packs
                val nativePacks = packIds.mapNotNull { packId ->
                    credentialPackAdapter.getNativePack(packId)
                }

                if (nativePacks.isEmpty()) {
                    callback(Result.success(DcApiError(
                        message = "No valid credential packs found"
                    )))
                    return@launch
                }

                // Save packs to StorageManager so the DC API Activity can load them
                // This is required because the Activity runs in a separate process
                // and needs to access credentials via StorageManager
                for (pack in nativePacks) {
                    try {
                        pack.save(storageManager)
                        Log.d(TAG, "Saved pack ${pack.id()} to StorageManager")
                    } catch (e: Exception) {
                        Log.e(TAG, "Failed to save pack ${pack.id()} to StorageManager", e)
                    }
                }

                // Register with CredentialManager
                registry?.register(nativePacks)

                // Track registered credentials
                var registeredCount = 0
                for (packId in packIds) {
                    val credentials = credentialPackAdapter.getNativeCredentials(packId)
                    for (credential in credentials) {
                        val mdoc = credential.asMsoMdoc() ?: continue
                        val credentialId = credential.id()
                        val docType = mdoc.doctype()

                        registeredCredentials[credentialId] = RegisteredCredentialInfo(
                            credentialId = credentialId,
                            docType = docType,
                            isRegistered = true
                        )
                        registeredCount++
                    }
                }

                callback(Result.success(DcApiSuccess(
                    message = "Registered $registeredCount credentials with CredentialManager"
                )))

            } catch (e: Exception) {
                callback(Result.success(DcApiError(
                    message = "Failed to register credentials: ${e.localizedMessage}"
                )))
            }
        }
    }

    override fun unregisterCredentials(
        credentialIds: List<String>,
        callback: (Result<DcApiResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                var unregisteredCount = 0

                for (credentialId in credentialIds) {
                    // Remove from tracking
                    registeredCredentials.remove(credentialId)
                    unregisteredCount++
                }

                // Note: Android CredentialManager doesn't have a direct unregister API
                // The credentials are re-registered on each call to registerCredentials
                // So we just clear our tracking here

                callback(Result.success(DcApiSuccess(
                    message = "Unregistered $unregisteredCount credentials"
                )))

            } catch (e: Exception) {
                callback(Result.success(DcApiError(
                    message = "Failed to unregister credentials: ${e.localizedMessage}"
                )))
            }
        }
    }

    override fun getRegisteredCredentials(): List<RegisteredCredentialInfo> {
        return registeredCredentials.values.toList()
    }

    override fun isSupported(): Boolean {
        // DC API on Android requires API level 34 (Android 14) or higher
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE
    }
}

File: flutter/ios/Classes/DcApiAdapter.swift
import Foundation
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs

#if canImport(IdentityDocumentServices)
import IdentityDocumentServices
#endif

/// DC API Adapter for iOS
///
/// Handles registering mDoc credentials with iOS IdentityDocumentProvider (iOS 26+)
class DcApiAdapter: DcApi {

    private let credentialPackAdapter: CredentialPackAdapter
    private var registeredCredentials: [String: RegisteredCredentialInfo] = [:]
    private let lock = NSLock()

    init(credentialPackAdapter: CredentialPackAdapter) {
        self.credentialPackAdapter = credentialPackAdapter
    }

    func syncCredentialsToAppGroup(
        appGroupId: String,
        packIds: [String],
        completion: @escaping (Result<DcApiResult, Error>) -> Void
    ) {
        // Not needed on iOS - credentials are loaded directly via CredentialPack.loadAll()
        // from the App Group StorageManager by the Extension
        completion(.success(DcApiSuccess(
            message: "Sync not needed on iOS (Extension uses StorageManager directly)"
        )))
    }

    func registerCredentials(
        packIds: [String],
        walletName: String?,
        completion: @escaping (Result<DcApiResult, Error>) -> Void
    ) {
        // Note: walletName is currently only used on Android for DC API display
        // iOS uses the app name from the extension's Info.plist
        Task {
            do {
                var registeredCount = 0
                var errors: [String] = []

                for packId in packIds {
                    // Get the native CredentialPack and use SDK's registration method
                    guard let credentialPack = credentialPackAdapter.getNativePack(packId: packId) else {
                        errors.append("CredentialPack not found for packId: \(packId)")
                        continue
                    }

                    // Count mDL credentials before registration
                    let credentials = credentialPack.list()
                    let mdlCredentials = credentials.filter { credential in
                        credential.asMsoMdoc()?.doctype() == "org.iso.18013.5.1.mDL"
                    }
                    let mdlCount = mdlCredentials.count

                    if mdlCount == 0 {
                        continue
                    }

                    // Use SDK's method which handles notAuthorized gracefully
                    do {
                        try await credentialPack.registerUnregisteredIDProviderDocuments()
                        registeredCount += mdlCount

                        // Track registered credentials
                        for credential in credentials {
                            if let mdoc = credential.asMsoMdoc(), mdoc.doctype() == "org.iso.18013.5.1.mDL" {
                                lock.lock()
                                registeredCredentials[credential.id()] = RegisteredCredentialInfo(
                                    credentialId: credential.id(),
                                    docType: mdoc.doctype(),
                                    isRegistered: true
                                )
                                lock.unlock()
                            }
                        }
                    } catch {
                        let errorMsg = "Registration failed: \(error)"
                        errors.append(errorMsg)
                    }
                }

                if registeredCount > 0 {
                    completion(.success(DcApiSuccess(
                        message: "Registered \(registeredCount) credentials"
                    )))
                } else if !errors.isEmpty {
                    completion(.success(DcApiError(
                        message: "No credentials registered. Errors: \(errors.joined(separator: "; "))"
                    )))
                } else {
                    completion(.success(DcApiError(
                        message: "No mDL credentials found to register"
                    )))
                }

            } catch {
                completion(.success(DcApiError(
                    message: "Failed to register credentials: \(error.localizedDescription)"
                )))
            }
        }
    }

    func unregisterCredentials(
        credentialIds: [String],
        completion: @escaping (Result<DcApiResult, Error>) -> Void
    ) {
        Task {
            do {
                var unregisteredCount = 0

                for credentialId in credentialIds {
                    // Unregister from iOS IdentityDocumentProvider (iOS 26+)
                    if #available(iOS 26.0, *) {
                        #if canImport(IdentityDocumentServices)
                        let store = IdentityDocumentProviderRegistrationStore()
                        try await store.removeRegistration(forDocumentIdentifier: credentialId)
                        #endif
                    }

                    lock.lock()
                    registeredCredentials.removeValue(forKey: credentialId)
                    lock.unlock()

                    unregisteredCount += 1
                }

                completion(.success(DcApiSuccess(
                    message: "Unregistered \(unregisteredCount) credentials"
                )))

            } catch {
                completion(.success(DcApiError(
                    message: "Failed to unregister credentials: \(error.localizedDescription)"
                )))
            }
        }
    }

    func getRegisteredCredentials() throws -> [RegisteredCredentialInfo] {
        lock.lock()
        let credentials = Array(registeredCredentials.values)
        lock.unlock()
        return credentials
    }

    func isSupported() throws -> Bool {
        if #available(iOS 26.0, *) {
            return true
        }
        return false
    }
}

